lib/rack/test/uploaded_file.rb

```diff
@@ -30,6 +30,8 @@ def initialize(path, content_type = "text/plain", binary = false)
         @tempfile.set_encoding(Encoding::BINARY) if @tempfile.respond_to?(:set_encoding)
         @tempfile.binmode if binary
 
+        ObjectSpace.define_finalizer(self, self.class.finalize(@tempfile))
+
```

Comment:
Check this out mate, http://www.mikeperham.com/2010/02/24/the-trouble-with-ruby-finalizers/